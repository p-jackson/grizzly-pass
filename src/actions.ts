import { uniqueId } from "lodash";
import type { Project, Label, Status, TabId } from "./types";

export interface LoadDemoData {
  type: "LOAD_DEMO_DATA";
}
export const loadDemoData = (): LoadDemoData => ({
  type: "LOAD_DEMO_DATA",
});

export interface LoadProjectJsonSuccess {
  type: "LOAD_PROJECT_JSON_SUCCESS";
  data: {
    title?: string;
    projects: Project[];
    labels: Label[];
  };
}
export const loadProjectJsonSuccess = (data: {
  title?: string;
  projects: Project[];
  labels: Label[];
}): LoadProjectJsonSuccess => ({
  type: "LOAD_PROJECT_JSON_SUCCESS",
  data,
});

export interface LoadProjectJsonFailure {
  type: "LOAD_PROJECT_JSON_FAILURE";
  errorMessage: string | string[];
}
export const loadProjectJsonFailure = (
  errorMessage: string | string[],
): LoadProjectJsonFailure => ({
  type: "LOAD_PROJECT_JSON_FAILURE",
  errorMessage,
});

export interface LoadProject {
  type: "LOAD_PROJECT";
  project: Project;
}
export const loadProject = (
  title: string,
  person: string,
  time: string,
  progress: number,
  status: Status,
  labels: string[],
): LoadProject => ({
  type: "LOAD_PROJECT",
  project: {
    id: uniqueId(),
    title,
    person,
    time,
    progress,
    status,
    labels,
  },
});

export interface LoadLabel {
  type: "LOAD_LABEL";
  label: Label;
}
export const loadLabel = (title: string): LoadLabel => ({
  type: "LOAD_LABEL",
  label: {
    id: uniqueId(),
    title,
  },
});

export interface UpdateProject {
  type: "UPDATE_PROJECT";
  project: Project;
}
export const updateProject = (project: Project): UpdateProject => ({
  type: "UPDATE_PROJECT",
  project,
});

export interface SelectTab {
  type: "SELECT_TAB";
  tabId?: TabId;
}
export const selectTab = (tabId?: TabId): SelectTab => ({
  type: "SELECT_TAB",
  tabId,
});

export type Action =
  | LoadDemoData
  | LoadProjectJsonSuccess
  | LoadProjectJsonFailure
  | LoadProject
  | LoadLabel
  | UpdateProject
  | SelectTab;
