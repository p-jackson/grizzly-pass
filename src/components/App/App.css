.App {
  position: absolute;
  width: 100%;
  min-height: 100%;
}

.App-header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 50px;
  z-index: 1;
}

.App-sideMenu {
  position: fixed;
  top: 50px;
  left: 0;
  bottom: 0;
  width: 60px;
  z-index: 1;
}

.App-footer {
  position: fixed;
  left: 60px;
  right: 0;
  bottom: 0;
  background: #f2f3f4;
}

.App-content {
  display: flex;
  flex-direction: row;
  margin-top: 50px;
  margin-left: 60px;
  padding: 30px 30px 80px 30px;
}

.App-month {
  padding-right: 30px;
}

.App-month:last-child {
  padding-right: 0;
}

.App-monthTitle {
  font-size: 23px;
  font-weight: 300;
  padding-bottom: 10px;
}

.App-cardWrapper {
  padding-bottom: 10px;
}
