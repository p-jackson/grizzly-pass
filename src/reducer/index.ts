import moment from "moment";
import { combineReducers } from "redux";
import type { Action } from "../actions";
import { projects as demoProjects, labels as demoLabels } from "../demo-data";
import { generateLabelInfo } from "../labels";
import type { Label, Project, LabelInfo, TabId } from "../types";

export type { Label, Project };

export interface State {
  readonly title: string | null;
  readonly selectedTab: TabId | null;
  readonly projects: Project[];
  readonly labels: Label[];
  readonly errorMessage: string | string[] | null;
}

const reducer = combineReducers<State>({
  title,
  selectedTab,
  projects,
  labels,
  errorMessage,
});
export default reducer;

function title(title: string | null = null, action: Action): string | null {
  if (action.type === "LOAD_DEMO_DATA") return "Demo Dashboard";
  else return title;
}

function selectedTab(
  selectedTab: TabId | null = null,
  action: Action,
): TabId | null {
  if (action.type === "SELECT_TAB")
    return action.tabId === undefined ? null : action.tabId;
  return selectedTab;
}

function projects(projects: Project[] = [], action: Action): Project[] {
  switch (action.type) {
    case "LOAD_DEMO_DATA":
      return demoProjects;

    case "LOAD_PROJECT":
      return [...projects, action.project];

    case "UPDATE_PROJECT": {
      const project = action.project;
      const index = projects.findIndex(({ id }) => id === project.id);
      if (index === -1) return projects;
      else
        return [
          ...projects.slice(0, index),
          action.project,
          ...projects.slice(index + 1),
        ];
    }

    default:
      return projects;
  }
}

function labels(labels: Label[] = [], action: Action): Label[] {
  switch (action.type) {
    case "LOAD_DEMO_DATA":
      return demoLabels;
    case "LOAD_LABEL":
      return [...labels, action.label];
    default:
      return labels;
  }
}

function errorMessage(
  errorMessage: string | string[] | null = null,
): string | string[] | null {
  return errorMessage;
}

export function getTitle(state: State): string | undefined {
  return state.title === null ? undefined : state.title;
}

export function getProject(state: State, id: string): Project | undefined {
  return state.projects.find((project) => project.id === id);
}

export function getUsedLabels(state: State): string[] {
  const allLabelIds = state.projects.reduce(
    (memo, project) => [...memo, ...project.labels],
    [] as string[],
  );
  return state.labels
    .filter((label) => allLabelIds.indexOf(label.id) !== -1)
    .map((label) => label.id);
}

export function getLabelInfo(state: State, id: string): LabelInfo {
  return generateLabelInfo(state.labels)[id];
}

export function getProjectIdsByMonth(
  state: State,
): { month: string; projectIds: string[] }[] {
  const monthIndexes = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
  return monthIndexes
    .map((monthIndex) =>
      state.projects.filter(
        (project) => moment(project.time).month() === monthIndex,
      ),
    )
    .filter((month) => month.length > 0)
    .map((month) => ({
      month: moment(month[0].time).format("MMMM"),
      projectIds: month.map((project) => project.id),
    }));
}

export function getErrorMessage(state: State): string | string[] | undefined {
  return state.errorMessage === null ? undefined : state.errorMessage;
}

export function getSelectedTab(state: State): TabId | undefined {
  return state.selectedTab === null ? undefined : state.selectedTab;
}

export function getEditable(state: State): boolean {
  return state.selectedTab === "edit";
}
