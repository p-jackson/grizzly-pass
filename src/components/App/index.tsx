import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { loadProjectJsonSuccess, loadProjectJsonFailure } from "../../actions";
import { importFile } from "../../import";
import { getProjectIdsByMonth, State } from "../../reducer";
import { TabId, Project } from "../../types";
import Card from "../Card";
import Header from "../Header";
import Legend from "../Legend";
import SideMenu from "../SideMenu";
import "./App.css";

const importProject = (
  readFileAsText: (file: File) => Promise<string>,
  file: File
) => async (dispatch: Dispatch<State>) => {
  const asText = await readFileAsText(file);
  importFile(asText)
    .map(fileData => dispatch(loadProjectJsonSuccess(fileData)))
    .mapErr(errorMessage => dispatch(loadProjectJsonFailure(errorMessage)));
};

interface OwnProps {
  readFileAsText: (file: File) => Promise<string>;
}

interface AppProps {
  title?: string;
  errorMessage?: string | string[];
  selectedTab?: TabId;
  projectsByMonth: { month: string; projectIds: string[] }[];
  updateProject: (project: Project) => void;
  importFile: (file: File) => void;
  selectTab: (tabId?: TabId) => void;
}

export function AppPresentation({
  projectsByMonth,
  errorMessage,
  selectedTab,
  importFile
}: AppProps) {
  const months = projectsByMonth.map(({ month, projectIds }) => {
    const cards = projectIds.map(projectId => (
      <div className="App-cardWrapper" key={projectId}>
        <Card projectId={projectId} />
      </div>
    ));
    return (
      <div key={month} className="App-month">
        <h2 className="App-monthTitle">{month}</h2>
        {cards}
      </div>
    );
  });

  const hasProjects = !!projectsByMonth.length;

  return (
    <div
      className="App"
      onDrop={e => handleDrop(importFile, e)}
      onDragOver={handleDragOver}
    >
      <div className="App-sideMenu">
        <SideMenu />
      </div>
      <div className="App-header">
        <Header />
      </div>
      <div className="App-content">
        {errorMessage != null ? formatErrors(errorMessage) : months}
      </div>
      {hasProjects &&
      errorMessage === undefined && (
        <div className="App-footer">
          <Legend />
        </div>
      )}
    </div>
  );
}

export function mapStateToProps(
  state: State,
  ownProps: OwnProps
): Partial<AppProps> {
  return {
    projectsByMonth: getProjectIdsByMonth(state)
  };
}

function mapDispatchToProps(
  dispatch: Dispatch<State>,
  { readFileAsText }: OwnProps
): Partial<AppProps> {
  return {
    importFile: (file: File) => dispatch(importProject(readFileAsText, file))
  };
}

const App = connect(mapStateToProps, mapDispatchToProps)(AppPresentation);
export default App;

function handleDragOver(e: React.SyntheticEvent<HTMLElement>) {
  e.preventDefault();
}

function handleDrop(
  importFile: (file: File) => void,
  e: React.DragEvent<HTMLElement>
) {
  e.preventDefault();
  const dt = e.dataTransfer;
  if (dt && dt.items) {
    const item = Array.prototype.find.call(
      dt.items,
      ({ kind }: DataTransferItem) => kind === "file"
    );
    if (item) importFile(item.getAsFile());
  } else if (dt && dt.files.length) {
    importFile(dt.files[0]);
  }
}

function formatErrors(errorMessage: string | string[]) {
  if (typeof errorMessage === "string") return errorMessage;

  return (
    <ul>{errorMessage.map((message, i) => <li key={i}>{message}</li>)}</ul>
  );
}
