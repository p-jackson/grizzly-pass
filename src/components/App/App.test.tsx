// @vitest-environment jsdom
import { fireEvent, render, screen } from "@testing-library/react";
import reducer from "../../reducer";
import { loadDemoData } from "../../actions";
import type { Project, TabId } from "../../types";
import { AppPresentation, mapStateToProps } from "../App";
import { createStore } from "redux";
import { Provider } from "react-redux";

const projectsByMonth: { month: string; projectIds: string[] }[] = [
  {
    month: "March",
    projectIds: ["1"],
  },
  {
    month: "April",
    projectIds: ["2"],
  },
];

function renderApp({
  title,
  projectsByMonth = [],
  importFile = vi.fn(),
  errorMessage,
  selectedTab,
  selectTab = vi.fn(),
  updateProject = vi.fn(),
}: {
  title?: string;
  projectsByMonth?: { month: string; projectIds: string[] }[];
  importFile?: (file: File) => void;
  errorMessage?: string | string[];
  selectedTab?: TabId;
  selectTab?: (tabId?: TabId) => void;
  updateProject?: (project: Project) => void;
} = {}) {
  const store = createStore(reducer);
  store.dispatch(loadDemoData());

  render(
    <Provider store={store}>
      <AppPresentation
        projectsByMonth={projectsByMonth}
        title={title}
        importFile={importFile}
        selectTab={selectTab}
        errorMessage={errorMessage}
        selectedTab={selectedTab}
        updateProject={updateProject}
      />
    </Provider>,
  );
}

it("shows a header", () => {
  renderApp();
  expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
    "Demo Dashboard",
  );
});

it("renders month names at the top of the columns", () => {
  renderApp({ projectsByMonth });
  const columnHeadings = screen.getAllByRole("heading", { level: 2 });
  expect(columnHeadings[0]).toHaveTextContent("March");
  expect(columnHeadings[1]).toHaveTextContent("April");
});

it("calls drop handler when browser support the DataTransferItemList interface", () => {
  const mockFile = { MOCK: "FILE" };
  const importFile = vi.fn();
  renderApp({ importFile });
  fireEvent.drop(screen.getByTestId("dropzone"), {
    preventDefault: () => {},
    dataTransfer: {
      items: [{ kind: "file", getAsFile: () => mockFile }],
    },
  });
  expect(importFile).toHaveBeenCalledWith(mockFile);
});

it("calls drop handler when browser support the DataTransfer interface", () => {
  const mockFile = { MOCK: "FILE" };
  const importFile = vi.fn();
  renderApp({ importFile });
  fireEvent.drop(screen.getByTestId("dropzone"), {
    preventDefault: () => {},
    dataTransfer: {
      files: [mockFile],
    },
  });
  expect(importFile).toHaveBeenCalledWith(mockFile);
});

it("doesn't call drop handler it wasn't a file that was dropped", async () => {
  const importFile = vi.fn();
  renderApp({ importFile });
  fireEvent.drop(screen.getByTestId("dropzone"), {
    preventDefault: () => {},
    dataTransfer: {
      items: [{ kind: "string", getAsFile: () => {} }],
    },
  });
  expect(importFile).not.toHaveBeenCalled();
});

it("displays errorMessage if it's passed as a prop", () => {
  renderApp({ errorMessage: "Error Name" });
  expect(screen.getByText("Error Name")).toBeInTheDocument();
});

it("displays errors as a list if errorMessage prop is an array", () => {
  const errorMessage = ["Error 1", "Error 2"];
  renderApp({ errorMessage });
  const items = screen.getAllByRole("listitem");
  expect(items).toHaveLength(2);
  expect(items[0]).toHaveTextContent("Error 1");
  expect(items[1]).toHaveTextContent("Error 2");
});

it("can map default state to props", () => {
  const state = reducer(undefined as any, { type: "DUMMY_ACTION" });
  expect(mapStateToProps(state, { readFileAsText: vi.fn() })).toMatchSnapshot();
});
