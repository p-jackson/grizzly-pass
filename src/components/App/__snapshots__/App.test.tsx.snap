// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can map default state to props 1`] = `
Object {
  "projectsByMonth": Array [],
}
`;
